spectator-ext-stackdriver/src/main/java/com/netflix/spectator/stackdriver/StackdriverWriter.java

```diff
@@ -0,0 +1,760 @@
+   * Stackdriver type name. That is, it does not yet contain the customTypePrefix.
+   * This is for [assumed] convienence specifying the hints.
+   */
+  public void addCustomDescriptorHints(List<CustomDescriptorHint> hints) {
+    for (CustomDescriptorHint hint : hints) {
+      addLabelHints(hint.name, hint.labels);
+    }
+  }
+
+  /**
+   * A simpler form of addCustomDescriptorHints.
+   */
+  public void addLabelHints(String name, List<String> labels) {
+   if (!name.startsWith(customTypePrefix)) {
+     name = customTypePrefix + name;
+   }
+   List<String> hints = labelHints.get(name);
+   if (hints == null) {
+     hints = new ArrayList<String>(labels.size());
+     hints.addAll(labels);
+     labelHints.put(name, hints);
+   } else {
+     hints.addAll(labels);
+   }
+  }
+
+  /**
+   * Helper function to fetch an individual descriptor from Stackdriver.
+   * This is used to lookup whether or not a particular type already exists.
+   * Normally we would already know this from initKnownDescriptors, however
+   * there is a race condition on first usage where the descriptor did not
+   * exist on our startup but another process has since created it.
+   * Fetching it will correct for that.
+   */
+    MetricDescriptor fetchDescriptorHelper(String descriptorName,
+                                           String descriptorType)
+       throws IOException {
+    MetricDescriptor descriptor
+        = service.projects().metricDescriptors().get(descriptorName).execute();
+    knownDescriptors.put(descriptorType, descriptor);
+    return descriptor;
+  }
+
+  /**
+   * Given a list of labels, maybe embellish it with additional labels
+   * depending on whether there are hints for the given descriptorType.
+   */
+  private void maybeAddLabelHints(String descriptorType,
+                                  List<LabelDescriptor> labels) {
+    List<String> ensure = labelHints.get(descriptorType);
+    if (ensure == null) return;
+
+    log.info("Verifying labels in {}", descriptorType);
+    for (String key : ensure) {
+      boolean found = false;
+      for (LabelDescriptor label : labels) {
+        if (label.getKey().equals(key)) {
+          log.info("Found hinted label '{}'", key);
+          found = true;
+          break;
+        }
+      }
+      if (! found) {
+        LabelDescriptor labelDescriptor = new LabelDescriptor();
+        log.info("Injecting hinted label '{}'.", key);
+        labelDescriptor.setKey(key);
+        labelDescriptor.setValueType("STRING");
+        labels.add(labelDescriptor);
+      }
+    }
+  }
+
+  /**
+   * Helper function for logging time series errors in more detail.
+   */
+  private void handleTimeSeriesResponseException(
+       HttpResponseException rex, String msg, List<TimeSeries> nextN) {
+    TimeSeries ts = findProblematicTimeSeriesElement(rex.getMessage(), nextN);
+    if (ts != null) {
+      log.error("{}\nTime Series element: {}",
```

Comment:
Generally I dislike line feeds in log statements as it makes it more difficult to reliably grep the log files.